free5gc-operator/controllers/smfdeployment_controller.go

```diff
@@ -0,0 +1,648 @@
+		                    dnnUpfInfoList: # DNN information list for this S-NSSAI
+		                      - dnn: internet
+		                        pools:
+		                          - cidr: 10.1.128.0/17
+		            interfaces: # Interface list for this UPF
+		                  - interfaceType: N3 # the type of the interface (N3 or N9)
+		                    endpoints: # the IP address of this N3/N9 interface on this UPF
+		                      - 10.100.50.233
+		                    networkInstance: internet # Data Network Name (DNN)
+		        links: # the topology graph of userplane, A and B represent the two nodes of each link
+		          - A: gNB1
+		            B: UPF
+		      locality: area1 # Name of the location where a set of AMF, SMF and UPFs are located
+
+		    logger:
+		      Aper:
+		        ReportCaller: false
+		        debugLevel: info
+		      NAS:
+		        ReportCaller: false
+		        debugLevel: info
+		      NGAP:
+		        ReportCaller: false
+		        debugLevel: info
+		      PFCP:
+		        ReportCaller: false
+		        debugLevel: info
+		      SMF:
+		        ReportCaller: false
+		        debugLevel: info
+	*/
+	smfcfgTemplate := template.New("SMFCfg")
+	smfcfgTemplate, err = smfcfgTemplate.Parse(SMFCfgTemplate)
+	if err != nil {
+		log.Log.Info("Could not parse SMFCfgTemplate template.")
+		return nil, err
+	}
+
+	smfueroutingTemplate := template.New("SMFCfg")
+	smfueroutingTemplate, _ = smfueroutingTemplate.Parse(Uerouting)
+	if err != nil {
+		log.Log.Info("Could not parse Uerouting template.")
+		return nil, err
+	}
+
+	var smfcfg bytes.Buffer
+	if err := smfcfgTemplate.Execute(&smfcfg, smfcfgStruct); err != nil {
+		log.Log.Info("Could not render SMFCfgTemplate template.")
+		return nil, err
+	}
+
+	var smf_uerouting bytes.Buffer
+	if err := smfueroutingTemplate.Execute(&smf_uerouting, smfcfgStruct); err != nil {
+		log.Log.Info("Could not render smf_uerouting template.")
+		return nil, err
+	}
+
+	configMap := &apiv1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ConfigMap",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instanceName + "-smf-configmap",
+			Namespace: namespace,
+		},
+		Data: map[string]string{
+			"smfcfg.yaml":    smfcfg.String(),
+			"uerouting.yaml": smf_uerouting.String(),
+		},
+	}
+	return configMap, nil
+}
+
+func (r *SMFDeploymentReconciler) syncSMFStatus(ctx context.Context, d *appsv1.Deployment, smfDeploy *workloadv1alpha1.SMFDeployment) error {
+	newSMFStatus, update := calculateSMFStatus(d, smfDeploy)
+
+	if update {
+		// Update SMFDeployment status according to underlying deployment status
+		newSmf := smfDeploy
+		newSmf.Status.NFDeploymentStatus = newSMFStatus
+		err := r.Status().Update(ctx, newSmf)
+		return err
+	}
+
+	return nil
+}
+
+func calculateSMFStatus(deployment *appsv1.Deployment, smfDeploy *workloadv1alpha1.SMFDeployment) (workloadv1alpha1.NFDeploymentStatus, bool) {
```

Comment:
We should have a `Ready` condition. Any error that occurs during processing should set the `Ready` condition to `False` and set the `Message` to the error string. This is how we surface errors to the user.

Well, there are basically two ways to surface errors. We should never expect the user to see a log file. Those are just for us, really. The resource status conditions is the primary way to surface an error *state*. However you can also fire *events*. Events in Kubernetes *are NOT for machine consumption*. They are only for users. They will show up, for example, when you `kubectl describe` the resource.

Currently I don't think we're firing events almost anywhere in Nephio. But eventually we should.

Reply:
Note that that link does NOT recommend Ready condition, but I think they are wrong.